drivers/clock_control/clock_control_rcar_cpg_mssr.c

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (c) 2020 IoT.bzh
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT renesas_rcar_cpg_mssr
+#include <errno.h>
+#include <soc.h>
+#include <drivers/clock_control.h>
+#include <drivers/clock_control/rcar_clock_control.h>
+#include <dt-bindings/clock/renesas_rcar_cpg.h>
+
+struct rcar_mssr_config {
+	uint32_t base_address;
+};
+
+#define DEV_CFG(dev)  ((struct rcar_mssr_config *)(dev->config))
+
+static const uint16_t rmstpsr[] = {
+	0x110, 0x114, 0x118, 0x11c,
+	0x120, 0x124, 0x128, 0x12c,
+	0x980, 0x984, 0x988, 0x98c,
+};
+
+#define RMSTPSR(i)      rmstpsr[i]
+
+/* Software Reset Register offsets */
+static const uint16_t srcr[] = {
+	0x0A0, 0x0A8, 0x0B0, 0x0B8,
+	0x0BC, 0x0C4, 0x1C8, 0x1CC,
+	0x920, 0x924, 0x928, 0x92C,
+};
+
+#define SRCR(i)         srcr[i]
+#define SRSTCLR(i)      (0x940 + (i) * 4)
+
+/* CPG write protect */
+#define CPGWPR                    0x0900
+/* CAN-FD Clock Frequency Control Register */
+#define CANFDCKCR                 0x244
+
+/* Clock stop bit */
+#define CANFDCKCR_CKSTP           BIT(8)
+
+/* On H3,M3,E3 parent clock of CANFD has 800MHz rate */
+#define CANFDCKCR_PARENT_CLK_RATE 800000000
+#define CANFDCKCR_DIVIDER_MASK    0x1FF
+
+#define S3D4_CLK_RATE             66600000
+
+static void cpg_write(uint32_t addr, uint32_t val)
+{
+	sys_write32(~val, CPGWPR);
+	sys_write32(val, addr);
+	/* Wait for at least one cycle of the RCLK clock (@ ca. 32 kHz) */
+	k_sleep(K_USEC(35));
+}
+
+static void cpg_reset(const struct rcar_mssr_config *config,
+		      uint32_t reg, uint32_t bit)
+{
+	uint32_t addr = config->base_address + SRCR(reg);
+
+	cpg_write(addr, BIT(bit));
+
+	addr = config->base_address + SRSTCLR(reg);
+	cpg_write(addr, BIT(bit));
+}
+
+static int cpg_core_clock_endisable(const struct device *dev,
+				    uint32_t module, uint32_t rate, bool enable)
+{
+	const struct rcar_mssr_config *config = DEV_CFG(dev);
+	uint32_t addr, divider;
+	int key, ret;
+
+	/* Only support CANFD core clock at the moment */
+	if (module != CPG_CORE_CLK_CANFD) {
+		return -EINVAL;
+	}
+
+	key = irq_lock();
+
+	addr = config->base_address + CANFDCKCR;
+	if (enable) {
+		if (CANFDCKCR_PARENT_CLK_RATE % rate) {
+			__ASSERT(true, "Can not generate "
+				 "%u from CANFD parent clock", rate);
+			ret = -EINVAL;
+			goto unlock;
+		}
+		divider = (CANFDCKCR_PARENT_CLK_RATE / rate) - 1;
+		if (divider > CANFDCKCR_DIVIDER_MASK) {
+			__ASSERT(true, "Can not generate %u "
+				 "from CANFD parent clock", rate);
+			ret = -EINVAL;
+			goto unlock;
+		}
+		cpg_write(addr, divider);
+	} else {
+		cpg_write(addr, CANFDCKCR_CKSTP);
+	}
+
+unlock:
+	irq_unlock(key);
+	return 0;
+}
+
+static int cpg_rmstp_clock_endisable(const struct device *dev,
+				     uint32_t module, bool enable)
+{
+	const struct rcar_mssr_config *config = DEV_CFG(dev);
+	uint32_t reg = module / 100;
+	uint32_t bit = module % 100;
+	uint32_t bitmask = BIT(bit);
```

Comment:
This seems somewhat unusual. Just double-checking that `module` never exceeds 31 for every 100 interval (i.e. 0-31, 100-131, 200-231, ...).

Reply:
ok will add a check

Reply:
fixed